idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirReferenceShortener.kt

```diff
@@ -295,60 +323,71 @@ private class ElementsToShortenCollector(
+        val allClassIds: Sequence<ClassId> = wholeClassQualifier.outerClassesWithSelf
+        val allQualifiers: Sequence<KtDotQualifiedExpression> = wholeQualifierElement.qualifiersWithSelf
+        return findElementsToShorten(positionScopes, allClassIds, allQualifiers, ::ShortenQualifier, this::findFakePackageToShorten)
+    }
 
-        val allClassIds = wholeClassQualifier.outerClassesWithSelf
-        val allQualifiers = wholeQualifierElement.qualifiersWithSelf
-
-        for ((classId, qualifier) in allClassIds.zip(allQualifiers)) {
-            val firstFoundClass = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)?.classId
-
-            if (firstFoundClass == classId) {
-                return ShortenQualifier(qualifier)
+    private inline fun <E> findElementsToShorten(
+        positionScopes: List<FirScope>,
+        allClassIds: Sequence<ClassId>,
+        allElements: Sequence<E>,
+        createElementToShorten: (E, nameToImport: FqName?, importAllInParent: Boolean) -> ElementToShorten,
+        findFakePackageToShortenFn: (E) -> ElementToShorten?,
+        elementFilter: (E) -> Boolean = { true },
+    ): ElementToShorten? {
+
+        for ((classId, element) in allClassIds.zip(allElements)) {
+            if (!elementFilter(element)) return null
+            val option = classShortenOption(classId.toRegularClass() ?: return null)
+            if (option == ShortenOption.DO_NOT_SHORTEN) continue
+
+            // Find class with the same name that's already available in this file.
+            val availableClassifier = shorteningContext.findFirstClassifierInScopesByName(positionScopes, classId.shortClassName)
+
+            when {
+                availableClassifier == null -> {
+                    if (option == ShortenOption.SHORTEN_IF_ALREADY_IMPORTED) continue
+                    return createElementToShorten(
+                        element,
+                        classId.asSingleFqName(),
+                        option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                    )
+                }
+                availableClassifier.symbol == classId -> {
+                    return when {
+                        !availableClassifier.isFromPackageImport && !availableClassifier.isFromStarImport &&
+                                option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
+                            createElementToShorten(element, classId.asSingleFqName(), true)
+                        }
+                        else -> createElementToShorten(element, null, false)
+                    }
+                }
+                // Allow using star import to overwrite members implicitly imported by default.
+                availableClassifier.isFromDefaultImport && availableClassifier.isFromStarImport && option == ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT -> {
```

Comment:
From your other comment, I understand this case better. But I would like to know why the 2nd clause (`&& availableClassifier.isFromStarImport`) is present, and why we will only allow overwriting with a star import (`SHORTEN_AND_IMPORT_ALL_IN_PARENT`).

The default imports right now are _all_ star imports (https://kotlinlang.org/docs/packages.html#default-imports) so the presence of the 2nd clause doesn't make a difference. However, I don't see why we couldn't overwrite a single default import. Say for whatever reason `kotlin.collections.Collection` is a default import. If you have a class `myPackage.Collection`, you could still overwrite the default import with `import myPackage.Collection`, right?

Reply:
> The default imports right now are all star imports
No, this is not the case in the current implementation. For example, `kotlin.String` is explicitly imported (https://github.com/JetBrains/kotlin/blob/f8d50d585d8f0940f4da30b6838cccfb3ca523bf/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/scopes/impl/FirExplicitSimpleImportingScope.kt#L13) and a star import of a custom `String` won't overwrite the bult-in String. Instead, `kotlin.String` can only be overwritten with a explicit import.

Reply:
I don't see `kotlin.String` explicitly imported there?

Yes, those can't be overwritten with a star import. But if they can be overwritten with an explicit import, then maybe there should be `|| availableClassifier.isFromDefaultImport` in the next branch? i.e.,`(availableClassifier.isFromStarImport || availableClassifier.isFromPackageImport || availableClassifier.isFromDefaultImport) && option == ShortenOption.SHORTEN_AND_IMPORT ->`

Reply:
In addition to symbols imported by default, any star-imported or package-imported symbols can be overwritten by explicit import. From my observation of FE1.0, the priority is
`explicit import = default explicit import > package import > star import > default star import`.